### PR TITLE
Enable loss scale input from Python frontend

### PR DESCRIPTION
**Pull request by edgchen1**
Made some fixes to enable loss scale to be wired up to ORT from the Python frontend. In particular, now addition of loss scaling is done unconditionally if mixed precision is enabled. The generated loss scale input name is passed back to the frontend.

Also fixed how inputs were added during the training graph configuration. Graph::SetInputs() was causing some issues - it seems to not be working correctly.

### CODE REVIEW

File: orttraining/orttraining/core/session/training_session.h [hunk 1]
```diff
@@ -215,19 +211,23 @@ class TrainingSession : public InferenceSession {
...
   */
-  common::Status BuildLossScalingFactorInput(const float loss_scale, std::string& loss_scale_input_name);
-
-  /** Add a system provided or an op as loss function to the model.
-  After the call, the model have one more input named as label_name and one more output named as loss_func_output_name.
-  @param loss_func_info The loss function info.
-  @param loss_scale_input_name If non-empty, specifies that loss scaling should be applied using the named
-         loss scaling factor input. Otherwise, loss scaling will not be applied.
-  @param actual_loss_name The actual name of the loss function output from which to start the backward graph.
+  common::Status BuildLossScalingFactorInput(std::string& loss_scale_input_name);
```
Comment (edgchen1) [inline orttraining/orttraining/core/session/training_session.h]: delete

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -617,11 +595,7 @@ def _init_session(self):
         if self.onnx_model_ is None:
             return
 
-        if self.use_mixed_precision:
-            self.loss_scale_input_name, self.scaled_loss_output_name = add_loss_scale_input(self.onnx_model_)
-            self.input_desc_with_lr_and_loss_scale = [*self.input_desc_with_lr, IODescription(self.loss_scale_input_name, [], torch.float32)]
-        else:
-            self.loss_scale_input_name, self.scaled_loss_output_name = '', ''
+        self.enable_grad_norm_clip_ = enable_grad_norm_clip
```
Comment (edgchen1) [inline orttraining/orttraining/python/ort_trainer.py]: merge issue?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 3]
```diff
@@ -182,6 +188,9 @@ void addObjectMethodsForTraining(py::module& m) {
       .def_readwrite("partition_optimizer", &TrainingParameters::partition_optimizer)
       .def_readwrite("enable_grad_norm_clip", &TrainingParameters::enable_grad_norm_clip);
 
+  py::class_<TrainingConfigurationResult> config_result(m, "TrainingConfigurationResult", "pbdoc(Configuration result for training.)pbdoc");
+  config_result.def(py::init())
+      .def_readonly("loss_scale_input_name", &TrainingConfigurationResult::loss_scale_input_name);
```
Comment (edgchen1) [inline orttraining/orttraining/python/orttraining_pybind_state.cc]: @liqun is it possible to have this return None if mixed precision is not enabled? wasn't sure how to do that

File: orttraining/orttraining/core/session/training_session.cc [hunk 4]
```diff
@@ -258,29 +251,58 @@ Status TrainingSession::ConfigureForTraining(
...
+    const std::string& loss_name,
+    Graph& graph, std::string* loss_scale_input_name, std::string& scaled_loss_name) {
+  if (!loss_scale_input_name) {
+    scaled_loss_name = loss_name;
+    return Status::OK();
+  }
+
+  // add node to scale loss_name by loss_scale_input_name
+  GraphAugmenter::GraphDefs defs{};
+  *loss_scale_input_name = graph.GenerateNodeArgName("loss_scale");
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/core/session/training_session.cc]: Do we need to `delete loss_scale_input_name` before allocating a new string to it?

File: orttraining/orttraining/core/session/training_session.cc [hunk 5]
```diff
@@ -123,42 +127,29 @@ Status TrainingSession::ConfigureForTraining(
...
-
-    if (mixed_precision_config.add_loss_scaling) {
-      ORT_RETURN_IF_ERROR(BuildLossScalingFactorInput(
-          mixed_precision_config.initial_loss_scale_value,
-          loss_scale_input_name));
-
-      tensorboard_scalar_names.emplace_back(loss_scale_input_name);
-      mixed_precision_config_result.loss_scale_input_name = loss_scale_input_name;
-    }
+  is_mixed_precision_enabled_ = config.mixed_precision_config.has_value();
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply orttraining/orttraining/core/session/training_session.cc]: `config.mixed_precision_config` is a `optional<TrainingSession::TrainingConfiguration::MixedPrecisionConfiguration>`. it's presence means that mixed precision is enabled

File: orttraining/orttraining/core/session/training_session.cc [hunk 6]
```diff
@@ -63,7 +63,9 @@ Status SetupOptimizerParams(
       return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, ex.what());
     }
 
-    opt_node_config.loss_scale_input_name = loss_scale_input_name;
+    // TODO make OptimizerNodeConfig::loss_scale_input_name optional<string>
+    opt_node_config.loss_scale_input_name =
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply orttraining/orttraining/core/session/training_session.cc]: it would be nice... unfortunately we have a limited optional which doesn't have value_or implemented.

[the remark this answers is not in the document]
Comment (edgchen1) [reply orttraining/orttraining/core/session/training_session.cc]: if you have a strong preference, i can try to implement it

[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/core/session/training_session.cc]: If it is not available, it is not needed

File: orttraining/orttraining/core/session/training_session.cc (hunk 4 above)
Comment (edgchen1) [reply orttraining/orttraining/core/session/training_session.cc]: this just assigns to the pointed-to string (checked to be not null earlier). a new string is not dynamically allocated here.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 3 above)
Comment (edgchen1) [reply orttraining/orttraining/python/orttraining_pybind_state.cc]: my understanding is that this code sets up the binding between C++ and Python constructs - so at this point we don't know whether mixed precision is enabled or not. let me know if i'm missing something

**Comment by edgchen1** [comment]
> I will look into it one more time. Is there any test for the loss scale input from python frontend?

Thanks for taking a look. @liqunfu do we have any python frontend tests that cover the loss scale input?

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 3 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/orttraining_pybind_state.cc]: You are right, it's been sometime since I worked with pybind, but you can use [lambdas](https://pybind11.readthedocs.io/en/stable/classes.html#binding-lambda-functions)


`.def_readonly('loss_scale_input_name', [](const TrainingConfigurationResult &res) {     
    if (res.mixed_precision_config_result.has_value()) {
        return res.loss_scale_input_name;
    } else {
        return None;
}`

That is a similar test from orttraining/orttraining/python/orttraining_pybind_state.cc, line 161

However, if we could use nullptr on C++ instead of "" or std::string(){}, this translation would be automatic in the current binding

Comment (liqunfu) [reply orttraining/orttraining/python/orttraining_pybind_state.cc]: Agreed. Probably None is better than an empty string?

**Comment by liqunfu** [comment]
@edgchen1 regarding the test, I have 2 mixed precision tests which are blocked by not having the CI machine able to run mixed precision models.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 3 above)
Comment (edgchen1) [reply orttraining/orttraining/python/orttraining_pybind_state.cc]: ok, figured out how to do that. thanks for the input

File: orttraining/orttraining/python/ort_trainer.py [hunk 7]
```diff
@@ -636,10 +608,16 @@ def _init_session(self):
...
 
+        self.loss_scale_input_name = self.session.loss_scale_input_name
+
+        if self.use_mixed_precision:
+            self.input_desc_with_lr_and_loss_scale = [
+                *self.input_desc_with_lr,
+                IODescription(self.loss_scale_input_name, [], torch.float32)]
+
         # ORT backend has modified model output dtype from float32 to float16.
         for o_desc in self.model_desc_.outputs_:
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply orttraining/orttraining/python/ort_trainer.py]: name of output loss also doesn't match right? for fp16 loss output will have generated name.

File: orttraining/orttraining/core/session/training_session.cc [hunk 8]
```diff
@@ -258,29 +251,58 @@ Status TrainingSession::ConfigureForTraining(
...
+      defs.CreateTypeProto({1}, ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
+  scaled_loss_name = graph.GenerateNodeArgName("scaled_loss");
+  defs.AddNodeDef(NodeDef{
+      "Mul",
+      {ArgDef{loss_name}, ArgDef{*loss_scale_input_name, loss_scale_input_type}},
+      {ArgDef{scaled_loss_name}},
+      NodeAttributes(),
+      scaled_loss_name});
+
+  ORT_RETURN_IF_ERROR(GraphAugmenter::AugmentGraph(graph, defs));
```
Comment (BowenBao) [inline orttraining/orttraining/core/session/training_session.cc]: do we need to replace current loss with scaled loss in graph outputs?

Comment (edgchen1) [reply orttraining/orttraining/core/session/training_session.cc]: i don't believe so. the scaled loss is used to build the backward graph, not aware of needing it as a graph output.

**Comment by SherlockNoMad** [comment]
What's the status of this PR? Is this ready to be merged?

**Comment by BowenBao** [comment]
Quick question, if I understand correctly, after this PR loss scaler input name will be generated from backend and returned to the frontend. Meanwhile learning rate input name will still be provided from user?

File: onnxruntime/test/python/onnxruntime_test_ort_trainer.py [hunk 9]
```diff
@@ -106,7 +106,7 @@ def runBertTrainingTest(gradient_accumulation_steps, use_mixed_precision, allred
 
         learning_rate = torch.tensor([lr]).to(device)
         if use_mixed_precision:
-            loss_scale = torch.tensor(loss_scaler.loss_scale_).to(device)
+            loss_scale = torch.tensor([loss_scaler.loss_scale_]).to(device)
```
Comment (BowenBao) [inline onnxruntime/test/python/onnxruntime_test_ort_trainer.py:109]: We are changing loss_scale from scalar(shape []) to 1-d tensor(shape [1]), is this change expected?

**Comment by edgchen1** [comment]
> Quick question, if I understand correctly, after this PR loss scaler input name will be generated from backend and returned to the frontend. Meanwhile learning rate input name will still be provided from user?

Yes

File: onnxruntime/test/python/onnxruntime_test_ort_trainer.py (hunk 9 above)
Comment (edgchen1) [reply onnxruntime/test/python/onnxruntime_test_ort_trainer.py:109]: Yep, it's consistent with the learning rate input.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 10]
```diff
@@ -157,6 +157,14 @@ void ConfigureSessionForTraining(
   training::TrainingSession::TrainingConfigurationResult config_result{};
 
   OrtPybindThrowIfError(sess->ConfigureForTraining(config, config_result));
+
+  TrainingConfigurationResult python_config_result{};
+  if (config_result.mixed_precision_config_result.has_value()) {
+    const auto& mp_config_result = config_result.mixed_precision_config_result.value();
+    python_config_result.loss_scale_input_name = mp_config_result.loss_scale_input_name;
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:164]: added assertion at the point where it's set in training_session.cc
